contributors/design-proposals/scheduling/scheduling-policy.md

```diff
@@ -0,0 +1,516 @@
+Then the most restrictive one (i.e the all bindingMode) is applied.
+
+### unset fields:
+
+if a field was not specified in the policy no rule should apply (i.e. everything is allowed).
+
+### empty match:
+
+Usually policies should distinguish between empty matches, since it depends on the action:
+
+- require/deny: no rule apply.
+- allow: allows everything.
+
+
+### inside the matches:
+
+The match field can express pretty much any structure you want to, but there's some things that should be considered:
+
+- the structure should match as much as possible what you try to take action on.
+- match elements are ANDed.
+- the `*` wildcard is usually needed in the allow section. It should be represented with an empty value (e.g. empty array).
+
+
+### conflict handling for policies
+
+There is two kinds of conflict in policies:
+
+- structural conflicts: these exists when any rule has the same matchingExpression and opposed actions (deny and require)
+
+- semantical conflicts: these exists due to semantic of the fields (e.g. require a NodeSelector `master=true` and also require `master=false`)
+
+
+Structural conflicts must handled at creation time, on the other hand, semantical conflicts should be handled at runtime: detect that there's a conflict and emit an event stating that policies couldn't be satisfied due to a conflict.
+
+
+### required vs allowed vs denied
+
+Policies may have overlapping rules, to handle this policies are computed in the following order:
+
+- compute what is denied.
+- compute what is required.
+- compute what was allowed.
+
+they should also obey to the following rules:
+
+- everything that is required is by definition allowed.
+- everything that is not denied is not automatically allowed.
+
+
+# Detailed Design
+
+
+## SchedulingPolicy
+
+Proposed API group: `policy/v1alpha1`
+
+
+### SchedulingPolicy content
+
+SchedulingPolicy spec is composed of optional fields that allow scheduling rules. If a field is absent from a SchedulingPolicy, this `SchedulingPolicy` won't allow any item from the missing field.
+
+```yaml
+apiVersion: policy/v1alpha1
+kind: SchedulingPolicy
+metadata:
+  name: mySchedulingPolicy
+spec:
+  bindingMode: all
+  namespaces:
+    - default
+  rules:
+    schedulerNames:
+      - match: []
+        action:
+    priorityClassNames:
+      - match: []
+        action:
+    tolerations:
+      - match: []
+        action:
+    nodeSelectors:
+      - match: []
+        action:
+    nodeAffinities:
+      - match: []
+        action:
+    podAntiAffinities:
+      - match: []
+        action:
+    podAffinities:
+      - match: []
+        action:
+```
+
+
+### required
+Elements here are required, pods won't schedule if they aren't present. Also note that if something is required it is also allowed.
+
+### allowed
+Elements here are allowed, the policy allows the presence of these elements. From Pod's perspective, a pod can use one or N of the allowed items.
```

Comment:
I'm curious about the usefulness of `allowed`. I feel it's useless (or maybe I missed something). Both `required` and `deny` have the semantics of terminating the evaluation of the policy as a decision is made. While `allowed` doesn't have the semantics of termination or deny. If the default (after all rules evaluated, none denied) is allow, then `allowed` rule is useless. Otherwise, it seems `deny` is useless.

Reply:
A sorta analogy: `required` is like a manager saying, _only these specific N people are allowed to come to this important meeting, no exceptions_. Whereas `allowed` is like when it's up to each manager to invite whoever to join, manager A can say  _these specific N people are allowed to go_ and manager B can say  _these other specific M people are allowed to go_ and neither cares that the other has invited people as well.

Reply:
`denied` would be when the boss says _these N people are banned from this meeting_ overriding any invitation they might have gotten.

Reply:
I see. My understanding is:

`required` => one_of
`allowed` => any_of
`deny` => none_of

Reply:
It is more like:

required => all of the specified properties must exist
allowed => zero or more of the specified properties can exist and nothing else is allowed
deny => none of the specified properties can exist

Reply:
Same here. Please make sure that the document is clear about the meaning of each of the actions.